pandas/core/sparse/array.py

```diff
@@ -704,41 +1035,157 @@ def mean(self, axis=0, *args, **kwargs):
             nsparse = self.sp_index.ngaps
             return (sp_sum + self.fill_value * nsparse) / (ct + nsparse)
 
-    def value_counts(self, dropna=True):
-        """
-        Returns a Series containing counts of unique values.
+    def transpose(self, *axes):
+        """Returns the SparseArray."""
+        return self
 
-        Parameters
-        ----------
-        dropna : boolean, default True
-            Don't include counts of NaN, even if NaN is in sp_values.
+    @property
+    def T(self):
+        """Returns the SparseArray."""
+        return self
+
+    # ------------------------------------------------------------------------
+    # Ufuncs
+    # ------------------------------------------------------------------------
+    def __abs__(self):
+        return np.abs(self)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        new_inputs = []
+        new_fill_values = []
+
+        special = {'add', 'sub', 'mul', 'pow', 'mod', 'floordiv', 'truediv',
+                   'divmod', 'eq', 'ne', 'lt', 'gt', 'le', 'ge', 'remainder'}
+        if compat.PY2:
+            special.add('div')
+        aliases = {
+            'subtract': 'sub',
+            'multiply': 'mul',
+            'floor_divide': 'floordiv',
+            'true_divide': 'truediv',
+            'power': 'pow',
+            'remainder': 'mod',
+            'divide': 'div',
+        }
+        op_name = ufunc.__name__
+        op_name = aliases.get(op_name, op_name)
+
+        if op_name in special:
+            if isinstance(inputs[0], type(self)):
+                # this is surely incorrect...
+                return getattr(self, '__{}__'.format(op_name))(inputs[1])
+            else:
+                return getattr(self, '__r{}__'.format(op_name))(inputs[0])
 
-        Returns
-        -------
-        counts : Series
-        """
-        keys, counts = algos._value_counts_arraylike(self.sp_values,
-                                                     dropna=dropna)
-        fcounts = self.sp_index.ngaps
-        if fcounts > 0:
-            if self._null_fill_value and dropna:
-                pass
+        for input in inputs:
+            if isinstance(input, type(self)):
+                new_inputs.append(self.sp_values)
+                new_fill_values.append(self.fill_value)
             else:
-                if self._null_fill_value:
-                    mask = pd.isna(keys)
-                else:
-                    mask = keys == self.fill_value
+                new_inputs.append(input)
```

Comment:
Handling of non-spase inputs is currently broken. This will fail for binop ufuncs.


```python
In [20]: np.gcd(a, b)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-20-ff216ef2b6bf> in <module>()
----> 1 np.gcd(a, b)

~/sandbox/pandas/pandas/core/sparse/array.py in __array_ufunc__(self, ufunc, method, *inputs, **kwargs)
   1088                 new_fill_values.append(input)
   1089
-> 1090         new_values = ufunc(*new_inputs, **kwargs)
   1091         new_fill = ufunc(*new_fill_values, **kwargs)
   1092         return type(self)(new_values, sparse_index=self.sp_index,

ValueError: operands could not be broadcast together with shapes (2,) (3,)
```

What should happen here... is unclear. It's assumed that the operation doesn't change the sparsity of the data, since we just return a SparseArray with `self.sp_index`.

Reply:
And to be clear, something like the following does change the sparsity:

```python
In [4]: a = pd.SparseArray([1, 2, 0, 0])

In [5]: b = np.array([1, 2, 3, 4])

In [6]: np.gcd(np.asarray(a), np.asarray(b))
Out[6]: array([1, 2, 3, 4])

In [7]: np.gcd(a, b)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-ff216ef2b6bf> in <module>()
----> 1 np.gcd(a, b)

~/sandbox/pandas/pandas/core/sparse/array.py in __array_ufunc__(self, ufunc, method, *inputs, **kwargs)
   1092                 new_fill_values.append(input)
   1093
-> 1094         new_values = ufunc(*new_inputs, **kwargs)
   1095         new_fill = ufunc(*new_fill_values, **kwargs)
   1096         return type(self)(new_values,

ValueError: operands could not be broadcast together with shapes (2,) (4,)
```

For now, I'm OK with those being unsupported...